### PR TITLE
fix 3D animation trembling

### PR DESCRIPTION
**Pull request by vcloarec**
3D animation often trembling wen camera move.
This PR fixes this issue.
I wanted to do a before/after, but too tricky to capture...

**EDIT:**
Finally...
**_Before:_**
before

**_After_**
after

### CODE REVIEW

File: src/app/3d/qgs3dnavigationwidget.cpp [hunk 1]
```diff
@@ -202,5 +202,5 @@ Qgs3DNavigationWidget::Qgs3DNavigationWidget( Qgs3DMapCanvas *parent ) : QWidget
 void Qgs3DNavigationWidget::updateFromCamera()
 {
   // Make sure the angle is between 0 - 359
-  mCompas->setValue( ( int( mParent3DMapCanvas->cameraController()->yaw() ) % 360 + 360 ) % 360 );
+  mCompas->setValue( fmod( fmod( mParent3DMapCanvas->cameraController()->yaw(), 360 ) + 360, 360 ) );
```
Comment (nyalldawson) [inline src/app/3d/qgs3dnavigationwidget.cpp:205]: This looks a little strange to me (the before one!) -- can you see why there's two mods here in the first place? Wouldn't 
 
    mCompas->setValue( fmod( mParent3DMapCanvas->cameraController()->yaw() + 360, 360 );

give the same result?

**Comment by nyalldawson** [comment]
Nice catch!

File: src/app/3d/qgs3dnavigationwidget.cpp (hunk 1 above)
Comment (vcloarec) [reply src/app/3d/qgs3dnavigationwidget.cpp:205]: for negative value I think.

Comment (nyalldawson) [reply src/app/3d/qgs3dnavigationwidget.cpp:205]: But that shouldn't make a difference, because eg mod(-270 + 360, 360 ) = 90.

**Comment by vcloarec** [comment]
> Nice catch!

yes, that was a while this bug annoy me... but quite tricky to find.

File: src/app/3d/qgs3dnavigationwidget.cpp (hunk 1 above)
Comment (vcloarec) [reply src/app/3d/qgs3dnavigationwidget.cpp:205]: indeed...

**Comment by nirvn** [comment]
YESSSSS. Thanks for fixing this!

**Comment by nirvn** [comment]
@vcloarec , actually I might ha ve shown unwarranted excitement just now :) does this PR only fixes trembling camera when using the compass or also via keyboard/mouse navigation?

**Comment by vcloarec** [comment]
@nirvn , on my side I've got the issue with animation, for now I don't see any issue when using keyboard/mouse navigation. But the issue come from the camera position that pass thought the compass for animation and navigation, it could fix the possible trembling with navigation... Better way to know is to try (I couldn't replicate navigation trembling for now).

**Comment by wonder-sk** [comment]
Oh the trembling was caused by the compass widget? That sounds scary... is it because setValue() in compass was triggering another signal to update camera? If so - shall we update the value using whileBlocking()? `whileBlocking( mCompas )->setValue(...)` ?

By the way, there is another bug that causes trembling - but that only affects cases when camera is far from the the center of the scene (e.g. 50km) - https://bugreports.qt.io/browse/QTBUG-69463

**Comment by vcloarec** [comment]
@wonder-sk , indeed there was a cyclic connection...

**Review by wonder-sk — APPROVED** [review]

**Pull request merged**